### PR TITLE
Feature Manage Cities - Fourth Challenge

### PR DESCRIPTION
**Pull request by FranRossi**
TL;DR
Completes the entire section of Managing a City for the fourth Challenge.

Main section of cities with the correspondent table and pagination of 5 cities per page.

image

I used tailwind components to make everything more consistent, and Ajax with Jquery for the CRUD actions. 
For the views, I use blade components. I do have a "home" component called layout which enables to reuse the styling for future pages.

For the Update action, I used a separate form that has all the information that can be changed except the id.

image

The error handling it is not finished. I just show an alert message when an error happens, in the future I will change it to show flash messages.
For example when trying to edit the city London and change the name of the city , which must be unique, to one that already exist it shows this alert message.

image

Probably making shorter Pull Request in the future.

### CODE REVIEW

File: fourth-challenge/app/Http/Controllers/CityController.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        City::create($this->validateCity())->with('success', 'City Created!');
+        $cities = City::orderByDesc('id')->paginate($this->cityPerPage);
+        return view('components.table', [
+            'cities' => $cities
+        ]);
+    }
+
+    protected function validateCity(City $city = null): array
+    {
+        $city = $city ?: new City();
```
Comment (felipe-crosa) [inline fourth-challenge/app/Http/Controllers/CityController.php]: No esta mal, solo como comentario:
En php existe el operador ?? que su funcion es justamente evitar este tipo de terniary expressions. Lo que hace ese operador es si vos pones valor1 ?? valor2, si el valor 1 es null entonces te retorna el valor2, sino devuelve el valor1. Justo en este caso te vendria util ya que podrias hacer $city = $city ?? new City()

File: fourth-challenge/app/Http/Controllers/CityController.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            'name' => ['required', Rule::unique('cities', 'name')->ignore($city)],
+            'flights_arriving' => ['required', 'integer'],
+            'flights_departing' => ['required', 'integer']
+        ]);
+    }
+
+    public function sort(Request $request)
+    {
+        $column = $request->input('column');
+        $direction = $request->input('direction');
```
Comment (felipe-crosa-7S) [inline fourth-challenge/app/Http/Controllers/CityController.php:62]: Podrias validar el contenido de estos campos porque sino creo que si llegas a hacer un orderBy de una columna que no existe se te va a caer

Comment (felipe-crosa) [inline fourth-challenge/app/Http/Controllers/CityController.php]: Podrias validar el contenido de estos campos porque sino creo que si llegas a hacer un orderBy de una columna que no existe se te va a caer

File: fourth-challenge/app/Http/Controllers/CityController.php (hunk 1 above)
Comment (FranRossi) [reply fourth-challenge/app/Http/Controllers/CityController.php]: Gracias por el comentario! Voy a cambiarlo como me sugeriste para usar ese operador de php

File: fourth-challenge/app/Http/Controllers/CityController.php (hunk 2 above)
Comment (FranRossi) [reply fourth-challenge/app/Http/Controllers/CityController.php]: Si, tenés razón. Voy a agregar una validación sobre estos campos

File: fourth-challenge/tests/Feature/Cities/CityTest.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            'flights_departing' => 100,
+        ])
+            ->assertSessionHasErrors('flights_arriving');
+
+        $this->patch(action([CityController::class, 'update'], $city), [
+            'name' => 'Miami',
+            'flights_arriving' => 100,
+            'flights_departing' => '',
+        ])
+            ->assertSessionHasErrors('flights_departing');
```
Comment (felipe-crosa) [inline fourth-challenge/tests/Feature/Cities/CityTest.php]: Para este tipo de tests que lo que haces en fin es repetir la misma assertion con pequeños cambios en los body de los requests, phpunit tiene lo que son dataProviders (https://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers). Esta funcionalidad te permite escribir un solo test y despues especificar una lista de distintos escenarios con los que queres que se corra, te sirve para evitar duplicar codigo, te queda mas prolijo y en fin dejas solo una assertion por test que es mejor practica

File: fourth-challenge/app/Http/Controllers/CityController.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    protected function validateCity(City $city = null): array
+    {
+        $city = $city ?: new City();
+
+        return request()->validate([
+            'name' => ['required', Rule::unique('cities', 'name')->ignore($city)],
+            'flights_arriving' => ['required', 'integer'],
+            'flights_departing' => ['required', 'integer']
```
Comment (felipe-crosa) [inline fourth-challenge/app/Http/Controllers/CityController.php]: que pasa si te tira un negativo?

Comment (FranRossi) [reply fourth-challenge/app/Http/Controllers/CityController.php]: Bien, no lo había visto a eso, ya quedo solucionado

File: fourth-challenge/tests/Feature/Cities/CityTest.php (hunk 3 above)
Comment (FranRossi) [reply fourth-challenge/tests/Feature/Cities/CityTest.php]: Si, perfecto. Ya lo cambie y queda mucho mas prolijo

File: fourth-challenge/app/Http/Controllers/CityController.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        ]);
+    }
+
+    protected function validateCity(City $city = null): array
+    {
+        $city = $city ?? new City();
+
+        return request()->validate([
+            'name' => ['required', Rule::unique('cities', 'name')->ignore($city)],
+//            'flights_arriving' => ['required', 'integer', 'min:0', ],
```
Comment (Juan2418) [inline fourth-challenge/app/Http/Controllers/CityController.php]: These comments should be deleted

File: fourth-challenge/public/js/cities/form.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                $('#table-container').html(data);
+                const newPaginationHtml = $(data).find('#pagination-container').html();
+                $('#pagination-container').html(newPaginationHtml);
+
+                // Reset the form inputs
+                $('#new-city-name').val('');
+                // $('#flights-arriving').val('');
+                // $('#flights-departing').val('');
+            },
+            error: function(xhr, status, error) {
```
Comment (Juan2418) [inline fourth-challenge/public/js/cities/form.js:18]: As a general rule, I would not create any pr with todos if I can help it. Either resolve all the todos before the request or those you won't resolve now can be loaded to the management system of choice

File: fourth-challenge/public/js/cities/form.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        $.ajax({
+            type: 'POST',
+            url: 'cities',
+            data: formData,
+            success: function(data) {
+                $('#table-container').html(data);
+                const newPaginationHtml = $(data).find('#pagination-container').html();
+                $('#pagination-container').html(newPaginationHtml);
+
+                // Reset the form inputs
```
Comment (Juan2418) [inline fourth-challenge/public/js/cities/form.js]: Is this comment necessary? If the code is not clear enough that you need to mark the code section with a subtitle maybe consider extracting the block into an auxiliary function

File: fourth-challenge/public/js/cities/form.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+});
+
+
+
+// Edit city
+$(document).ready(function() {
+    $('.edit-form').on('submit', function(e) {
+        e.preventDefault();
+
+        var formData = $(this).serialize();
```
Comment (Juan2418) [inline fourth-challenge/public/js/cities/form.js]: Why `var`?

File: fourth-challenge/public/js/cities/table.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
+
+
+// delete city
+$(document).ready(function() {
+    $('.delete-form').on('submit', function(e) {
+        e.preventDefault();
```
Comment (Juan2418) [inline fourth-challenge/public/js/cities/table.js:4]: Same thing with the comments

File: fourth-challenge/resources/views/cities/edit-form.blade.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    <div class="px-4 sm:px-0">
+      <h2 class="text-base font-semibold leading-7 text-gray-900">Edit a city</h2>
+      <p class="mt-1 text-sm leading-6 text-gray-600">This information will be publicly display.</p>
+    </div>
+
+    <form class="edit-form bg-white shadow-sm ring-1 ring-gray-900/5 sm:rounded-xl md:col-span-1" id="{{$city->id}}">
+     @csrf
+      <div class="px-4 py-6 sm:p-8">
+        <div class="grid max-w-sm grid-cols-1 gap-y-8 gap-x-6 sm:grid-cols-1">
+{{--            TODO pasarlo a un componente--}}
```
Comment (Juan2418) [inline fourth-challenge/resources/views/cities/edit-form.blade.php]: Same thing here with the TODOs and please choose a language and stick with it, either everything in Spanish or everything in English. One or the other in this context is the same to me but consistency is important

File: fourth-challenge/resources/views/components/layout.blade.php [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+                            <a href="#" class="text-white hover:bg-indigo-500 hover:bg-opacity-75 rounded-md px-3 py-2 text-sm font-medium">Projects</a>
+
+                            <a href="#" class="text-white hover:bg-indigo-500 hover:bg-opacity-75 rounded-md px-3 py-2 text-sm font-medium">Calendar</a>
+
+                            <a href="#" class="text-white hover:bg-indigo-500 hover:bg-opacity-75 rounded-md px-3 py-2 text-sm font-medium">Reports</a>
+                        </div>
+                    </div>
+                </div>
+{{--                <div class="hidden md:block">--}}
```
Comment (Juan2418) [inline fourth-challenge/resources/views/components/layout.blade.php]: What is this huge comment about? Should it be there?

File: fourth-challenge/resources/views/components/table-header-sort.blade.php [hunk 12]
```diff
@@ -0,0 +1,12 @@
+@props(['name','currentColumn'])
+<th scope="col" class="px-3 py-3.5 text-left text-sm font-semibold text-gray-900 ">
+    <a href="#" class="sort group inline-flex" data-column="{{$currentColumn}}" data-direction="{{ $currentColumn ?? $currentDirection ?? 'asc' }}">
+        {{ $name}}
+        <!-- Active: "bg-gray-200 text-gray-900 group-hover:bg-gray-300", Not Active: "invisible text-gray-400 group-hover:visible group-focus:visible" -->
```
Comment (Juan2418) [inline fourth-challenge/resources/views/components/table-header-sort.blade.php]: Is this comment necessary?

File: fourth-challenge/routes/web.php [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+use App\Http\Controllers\CityController;
+use Illuminate\Support\Facades\Route;
+
+
+
+//Route::get('/', [CityController::class, 'index']);
```
Comment (Juan2418) [inline fourth-challenge/routes/web.php]: Shouldn't this be deleted?

**Review by Juan2418 — CHANGES_REQUESTED** [review]
Summing up, I know you already know this but, this PR is huge

File: fourth-challenge/app/Http/Controllers/CityController.php (hunk 5 above)
Comment (FranRossi) [reply fourth-challenge/app/Http/Controllers/CityController.php]: Right I was just trying to use accessors for this fields that Felipe suggested me earlier, and I forgot to delete these comments

**Comment by FranRossi** [comment]
> Summing up, I know you already know this but, this PR is huge

Thanks for all the comments! Yes I know it is huge, my bad won't happen again

File: fourth-challenge/public/js/cities/table.js [hunk 14]
```diff
@@ -28,26 +26,26 @@ $(document).ready(function () {
     console.log("Binding click event handler...");
     $('#table-container').on('click', '.sort', function (e) {
         e.preventDefault();
-        let column = $(this).data('column') || 'id';
-        let direction = $(this).data('direction') || 'asc';
-
+        $column = $(this).data('column');
+        $direction = $(this).data('direction') || 'asc';
+        console.log("Sorting by column: " + column + " in direction: " + direction);
```
Comment (Juan2418) [inline fourth-challenge/public/js/cities/table.js]: Unless this log is a requirement, please avoid unnecessary logs on the client

**Review by Juan2418 — CHANGES_REQUESTED** [review]
